reference/matrix/fbcsr_kernels.cpp

```diff
@@ -0,0 +1,486 @@
+    const ValueType *const bvals = source->get_const_values();
+
+    assert(nbrows * bs == result->get_size()[0]);
+    assert(nbcols * bs == result->get_size()[1]);
+    assert(source->get_num_stored_elements() ==
+           result->get_num_stored_elements());
+
+    IndexType *const row_ptrs = result->get_row_ptrs();
+    IndexType *const col_idxs = result->get_col_idxs();
+    ValueType *const vals = result->get_values();
+
+    const gko::blockutils::DenseBlocksView<const ValueType, IndexType> bvalues(
+        bvals, bs, bs);
+
+    for (IndexType brow = 0; brow < nbrows; ++brow) {
+        const IndexType nz_browstart = browptrs[brow] * bs * bs;
+        const IndexType numblocks_brow = browptrs[brow + 1] - browptrs[brow];
+        for (int ib = 0; ib < bs; ib++)
+            row_ptrs[brow * bs + ib] = nz_browstart + numblocks_brow * bs * ib;
+
+        for (IndexType ibnz = browptrs[brow]; ibnz < browptrs[brow + 1];
+             ++ibnz) {
+            const IndexType bcol = bcolinds[ibnz];
+
+            for (int ib = 0; ib < bs; ib++) {
+                const IndexType row = brow * bs + ib;
+                const IndexType inz_blockstart =
+                    row_ptrs[row] + (ibnz - browptrs[brow]) * bs;
+
+                for (int jb = 0; jb < bs; jb++) {
+                    const IndexType inz = inz_blockstart + jb;
+                    vals[inz] = bvalues(ibnz, ib, jb);
+                    col_idxs[inz] = bcol * bs + jb;
+                }
+            }
+        }
+    }
+
+    row_ptrs[source->get_size()[0]] =
+        static_cast<IndexType>(source->get_num_stored_elements());
+}
+
+GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
+    GKO_DECLARE_FBCSR_CONVERT_TO_CSR_KERNEL);
+
+
+template <typename ValueType, typename IndexType, typename UnaryOperator,
+          bool transpose_blocks>
+void convert_fbcsr_to_fbcsc(const IndexType num_blk_rows, const int blksz,
+                            const IndexType *const row_ptrs,
+                            const IndexType *const col_idxs,
+                            const ValueType *const fbcsr_vals,
+                            IndexType *const row_idxs,
+                            IndexType *const col_ptrs,
+                            ValueType *const csc_vals, UnaryOperator op)
+{
+    const gko::blockutils::DenseBlocksView<const ValueType, IndexType> rvalues(
+        fbcsr_vals, blksz, blksz);
+    gko::blockutils::DenseBlocksView<ValueType, IndexType> cvalues(
+        csc_vals, blksz, blksz);
+    for (IndexType brow = 0; brow < num_blk_rows; ++brow) {
+        for (auto i = row_ptrs[brow]; i < row_ptrs[brow + 1]; ++i) {
+            const auto dest_idx = col_ptrs[col_idxs[i]];
+            col_ptrs[col_idxs[i]]++;
+            row_idxs[dest_idx] = brow;
+            for (int ib = 0; ib < blksz; ib++) {
+                for (int jb = 0; jb < blksz; jb++) {
+                    cvalues(dest_idx, ib, jb) =
+                        op(transpose_blocks ? rvalues(i, jb, ib)
```

Comment:
Is there any case where you would not transpose the blocks ? Wouldn't that be an inequivalent matrix ?

Reply:
Yes, it would be a different matrix. There might be a use case for converting FBCSR to FBCSC without transposing the blocks necessarily. Right now though, this would not be possible to do from user code, but I could add a Fbcsr public function for this later.